File: android/app/src/main/java/com/rork/guidestreamtvandroid/data/remote/RemoteConfigService.kt
package com.rork.guidestreamtvandroid.data.remote

import android.content.Context
import com.rork.guidestreamtvandroid.data.remote.SupabaseManager.client
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.coroutines.CancellationException
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement

/**
 * Lightweight remote-configuration layer — mirrors iOS RemoteConfigService.swift.
 *
 * Reads AdMob ad unit IDs and Rakuten merchant IDs from `public.app_config`
 * via the existing Supabase client so they can be updated without shipping a
 * new build. Hydrates synchronously from a SharedPreferences cache on init
 * so values are available immediately at cold launch before the network
 * returns. `load()` never throws and silently no-ops on any failure, leaving
 * cached or hardcoded values intact.
 *
 * No secrets, tokens, or API keys are stored in or read from `app_config`.
 */
object RemoteConfigService {

    private const val CACHE_KEY = "gs_remote_config_cache"
    private const val PREFS_NAME = "gs_prefs"

    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

    @Volatile private var adsConfig: RemoteAdConfig? = null
    @Volatile private var rakutenConfig: RemoteRakutenConfig? = null
    @Volatile private var appContext: Context? = null

    @Serializable
    private data class RemoteAdConfig(
        val banner: String? = null,
        val interstitial: String? = null,
        val native: String? = null,
    )

    @Serializable
    private data class RemoteRakutenConfig(
        @SerialName("publisher_id") val publisherId: String? = null,
        val merchants: RemoteRakutenMerchants? = null,
    )

    @Serializable
    private data class RemoteRakutenMerchants(
        val netflix: String? = null,
        val hulu: String? = null,
        val disney: String? = null,
        val hbo: String? = null,
        val apple: String? = null,
        val peacock: String? = null,
        val paramount: String? = null,
        val prime: String? = null,
    )

    @Serializable
    private data class AppConfigRow(
        val key: String,
        val value: JsonElement,
    )

    /**
     * Synchronously hydrate from the SharedPreferences cache so values are
     * available before the network returns. Call from Application.onCreate()
     * before any ad or affiliate request fires. Also stashes the application
     * context so `load()` can write back to the cache.
     */
    fun init(context: Context) {
        appContext = context.applicationContext
        try {
            val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            val raw = prefs.getString(CACHE_KEY, null) ?: return
            val rows = json.decodeFromString<List<AppConfigRow>>(raw)
            ingest(rows, persist = false)
        } catch (_: Throwable) {
            // Corrupt cache — ignore; the network load will repopulate.
        }
    }

    /**
     * Reads all rows from `app_config`, decodes the three known keys, stores
     * them in memory, and persists the raw JSON to SharedPreferences. Never
     * throws — any failure leaves cached or hardcoded values intact.
     */
    suspend fun load() {
        try {
            val rows = client.postgrest
                .from("app_config")
                .select()
                .decodeList<AppConfigRow>()
            ingest(rows, persist = true)
        } catch (e: CancellationException) {
            throw e
        } catch (_: Throwable) {
            // Silent no-op: cached or hardcoded values stay in place.
        }
    }

    private fun ingest(rows: List<AppConfigRow>, persist: Boolean) {
        var newAds: RemoteAdConfig? = null
        var newRakuten: RemoteRakutenConfig? = null
        for (row in rows) {
            try {
                val data = json.encodeToString(JsonElement.serializer(), row.value)
                when (row.key) {
                    "ads_android" -> {
                        newAds = json.decodeFromString(RemoteAdConfig.serializer(), data)
                    }
                    "ads_ios" -> {
                        // Only used as a fallback when ads_android is missing.
                        if (newAds == null) {
                            newAds = json.decodeFromString(RemoteAdConfig.serializer(), data)
                        }
                    }
                    "affiliate_rakuten" -> {
                        newRakuten = json.decodeFromString(
                            RemoteRakutenConfig.serializer(),
                            data,
                        )
                    }
                }
            } catch (_: Throwable) {
                // Malformed row — skip; never crash.
            }
        }
        if (newAds != null) adsConfig = newAds
        if (newRakuten != null) rakutenConfig = newRakuten

        if (persist) {
            try {
                val encoded = json.encodeToString(
                    ListSerializer(AppConfigRow.serializer()),
                    rows,
                )
                appContext?.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                    ?.edit()?.putString(CACHE_KEY, encoded)?.apply()
            } catch (_: Throwable) {
                // Persistence failure is non-fatal.
            }
        }
    }

    /**
     * Returns the ad unit id for the given slot ("banner", "interstitial",
     * "native"), or null when the key is missing, null, or an empty string.
     */
    fun adUnit(slot: String): String? {
        val ads = adsConfig ?: return null
        val raw = when (slot) {
            "banner" -> ads.banner
            "interstitial" -> ads.interstitial
            "native" -> ads.native
            else -> null
        }
        return raw?.takeIf { it.isNotBlank() }
    }

    /** Returns the Rakuten publisher id, or null when missing/empty. */
    fun rakutenPublisherId(): String? =
        rakutenConfig?.publisherId?.takeIf { it.isNotBlank() }

    /**
     * Returns the Rakuten merchant id for the given service key (e.g.
     * "netflix", "hulu"), or null when missing/empty.
     */
    fun rakutenMerchantId(key: String): String? {
        val merchants = rakutenConfig?.merchants ?: return null
        val raw = when (key) {
            "netflix" -> merchants.netflix
            "hulu" -> merchants.hulu
            "disney" -> merchants.disney
            "hbo" -> merchants.hbo
            "apple" -> merchants.apple
            "peacock" -> merchants.peacock
            "paramount" -> merchants.paramount
            "prime" -> merchants.prime
            else -> null
        }
        return raw?.takeIf { it.isNotBlank() }
    }
}

File: ios/GuideStreamTV/Services/RemoteConfigService.swift
//
//  RemoteConfigService.swift
//  GuideStreamTV
//
//  Lightweight remote-configuration layer that lets AdMob ad unit IDs and
//  Rakuten merchant IDs be updated from Supabase (`public.app_config`)
//  without shipping a new build.
//
//  Strategy:
//  1. On init, synchronously hydrate from a UserDefaults cache so values
//     are available immediately at cold launch, before the network returns.
//  2. `load()` reads every row from `app_config`, decodes the three known
//     keys into typed structs, stores them in memory, and persists the raw
//     JSON back to UserDefaults so the next cold launch has a last-known-
//     good copy.
//  3. `load()` never throws to the caller and silently no-ops on any
//     network or decode failure, leaving cached or hardcoded values intact.
//
//  No secrets, tokens, or API keys are stored in or read from `app_config`.
//

import Foundation
import Supabase

/// Typed shape of an `ads_ios` / `ads_android` row value.
nonisolated struct RemoteAdConfig: Codable, Sendable {
    let banner: String?
    let interstitial: String?
    let native: String?
}

/// Typed shape of the `affiliate_rakuten` row value.
nonisolated struct RemoteRakutenConfig: Codable, Sendable {
    let publisherId: String?
    let merchants: RemoteRakutenMerchants?

    enum CodingKeys: String, CodingKey {
        case publisherId = "publisher_id"
        case merchants
    }
}

nonisolated struct RemoteRakutenMerchants: Codable, Sendable {
    let netflix: String?
    let hulu: String?
    let disney: String?
    let hbo: String?
    let apple: String?
    let peacock: String?
    let paramount: String?
    let prime: String?
}

/// Raw row shape returned by `select()` on `app_config`.
nonisolated struct AppConfigRow: Codable, Sendable {
    let key: String
    let value: AnyJSON
}

@MainActor
final class RemoteConfigService {
    static let shared = RemoteConfigService()

    /// UserDefaults key for the last-known-good raw JSON payload.
    private let cacheKey = "gs_remote_config_cache"

    /// In-memory decoded values. Populated from cache on init, refreshed by
    /// `load()`. All accessors read these optionals directly.
    private var adsConfig: RemoteAdConfig?
    private var rakutenConfig: RemoteRakutenConfig?

    private init() {
        hydrateFromCache()
    }

    // MARK: - Load

    /// Reads all rows from `app_config`, decodes the three known keys, stores
    /// them in memory, and persists the raw JSON to UserDefaults. Never
    /// throws — any failure leaves cached or hardcoded values intact.
    func load() async {
        do {
            let rows: [AppConfigRow] = try await SupabaseManager.shared.client
                .from("app_config")
                .select()
                .execute()
                .value
            ingest(rows)
        } catch {
            // Silent no-op: cached or hardcoded values stay in place.
        }
    }

    /// Decodes the supplied rows into the typed structs, then persists the
    /// raw JSON to UserDefaults so the next cold launch can hydrate from it.
    private func ingest(_ rows: [AppConfigRow]) {
        let encoder = JSONEncoder()
        let decoder = JSONDecoder()
        var newAds: RemoteAdConfig?
        var newRakuten: RemoteRakutenConfig?

        for row in rows {
            // Convert the AnyJSON value to Data via JSONEncoder (AnyJSON is
            // Encodable) so we can decode it into a typed struct.
            guard let data = try? encoder.encode(row.value) else { continue }
            switch row.key {
            case "ads_ios":
                newAds = try? decoder.decode(RemoteAdConfig.self, from: data)
            case "ads_android":
                // Only used as a fallback when ads_ios is missing.
                if newAds == nil {
                    newAds = try? decoder.decode(RemoteAdConfig.self, from: data)
                }
            case "affiliate_rakuten":
                newRakuten = try? decoder.decode(RemoteRakutenConfig.self, from: data)
            default:
                break
            }
        }

        if newAds != nil { adsConfig = newAds }
        if newRakuten != nil { rakutenConfig = newRakuten }

        // Persist the raw rows so the next cold launch can hydrate from cache.
        if let cacheData = try? encoder.encode(rows) {
            UserDefaults.standard.set(cacheData, forKey: cacheKey)
        }
    }

    // MARK: - Accessors

    /// Returns the ad unit id for the given slot ("banner", "interstitial",
    /// "native"), or nil when the key is missing, null, or an empty string.
    func adUnit(_ slot: String) -> String? {
        guard let ads = adsConfig else { return nil }
        let raw: String?
        switch slot {
        case "banner": raw = ads.banner
        case "interstitial": raw = ads.interstitial
        case "native": raw = ads.native
        default: raw = nil
        }
        guard let value = raw, !value.isEmpty else { return nil }
        return value
    }

    /// Returns the Rakuten publisher id, or nil when missing/empty.
    var rakutenPublisherId: String? {
        guard let id = rakutenConfig?.publisherId, !id.isEmpty else { return nil }
        return id
    }

    /// Returns the Rakuten merchant id for the given service key (e.g.
    /// "netflix", "hulu"), or nil when missing/empty.
    func rakutenMerchantId(for key: String) -> String? {
        guard let merchants = rakutenConfig?.merchants else { return nil }
        let raw: String?
        switch key {
        case "netflix": raw = merchants.netflix
        case "hulu": raw = merchants.hulu
        case "disney": raw = merchants.disney
        case "hbo": raw = merchants.hbo
        case "apple": raw = merchants.apple
        case "peacock": raw = merchants.peacock
        case "paramount": raw = merchants.paramount
        case "prime": raw = merchants.prime
        default: raw = nil
        }
        guard let value = raw, !value.isEmpty else { return nil }
        return value
    }

    // MARK: - Cache

    /// Synchronously hydrates the typed structs from the UserDefaults raw
    /// JSON cache so values are available before the network returns.
    private func hydrateFromCache() {
        guard let data = UserDefaults.standard.data(forKey: cacheKey) else { return }
        let decoder = JSONDecoder()
        let encoder = JSONEncoder()
        guard let rows = try? decoder.decode([AppConfigRow].self, from: data) else {
            return
        }
        for row in rows {
            guard let rowData = try? encoder.encode(row.value) else { continue }
            switch row.key {
            case "ads_ios":
                adsConfig = try? decoder.decode(RemoteAdConfig.self, from: rowData)
            case "ads_android":
                if adsConfig == nil {
                    adsConfig = try? decoder.decode(RemoteAdConfig.self, from: rowData)
                }
            case "affiliate_rakuten":
                rakutenConfig = try? decoder.decode(RemoteRakutenConfig.self, from: rowData)
            default:
                break
            }
        }
    }
}
